Forward / backward NJT support for several activation functions

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141604
* #141506
* #141392
* #140161
* __->__ #140736
* #141500

Several activation functions were unimplemented due to missing `pointwise` tags. This PR adds them and corresponding backwards implementations.

jbschlosser committed:
Forward / backward NJT support for several activation functions

[ghstack-poisoned]

Skylion007 commented on torch/nested/_internal/ops.py:
```diff
@@ -2212,6 +2212,41 @@ def new_empty_default(func, *args, **kwargs):
+)
+def activation_backward(func, *args, **kwargs):
+    # first NJT arg is expected to be grad_output
+    grad_output = next(arg for arg in args if isinstance(arg, NestedTensor))
```
Add an. assert that arg is not empty?

jbschlosser replied:
`check_schema()` handles this part :)

Skylion007 replied:
But couldn't none of the args be NestedTensors?

soulitzer replied:
We wouldn't have reached this torch dispatch if none are NestedTensors

jbschlosser replied:
`check_schema()` will validate that `self` is an NJT so we're guaranteed at least one, plus we're inside of `NestedTensor.__torch_dispatch__()` so we had to get there somehow from a `NestedTensor` instance

Skylion007 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5031,6 +5031,7 @@
   python_module: nn
+  tags: pointwise
 
 - func: relu6_(Tensor(a!) self) -> Tensor(a!)
```
The inplace versions should also be marked as pointwise, right?

Skylion007 replied:
Ah, but lots of other inplace ops are already tagged as pointwise unfortunately.

jbschlosser replied:
hm true, I think I'll open a follow-up to handle this properly

jbschlosser replied:
oh, this actually already works today:
```python
def test_inplace_unary_op_returns_same_instance(self, device, dtype):
    nt = random_nt_from_dims(
        [3, None, 5],
        device=device,
        dtype=dtype,
        layout=torch.jagged,
    )
    out = nt.relu_()
    # passes!
    self.assertIs(out, nt)
```

What I believe happens is that the ADInplaceOrView key handles aliasing.

jbschlosser committed:
Update on "Forward / backward NJT support for several activation functions"


Several activation functions were unimplemented due to missing `pointwise` tags. This PR adds them and corresponding backwards implementations.

[ghstack-poisoned]

jbschlosser committed:
Update on "Forward / backward NJT support for several activation functions"


Several activation functions were unimplemented due to missing `pointwise` tags. This PR adds them and corresponding backwards implementations.

[ghstack-poisoned]

jbschlosser committed:
Update on "Forward / backward NJT support for several activation functions"


Several activation functions were unimplemented due to missing `pointwise` tags. This PR adds them and corresponding backwards implementations.

[ghstack-poisoned]

jbschlosser committed:
Update on "Forward / backward NJT support for several activation functions"


Several activation functions were unimplemented due to missing `pointwise` tags. This PR adds them and corresponding backwards implementations.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but its tests are failing in trunk' -c nosignal

test_nestedtensor.py::TestNestedTensorOpInfoCUDA::test_compile_backward_nn_functional_softshrink_cuda_float32 [GH job link](https://github.com/pytorch/pytorch/actions/runs/11980859033/job/33407057580) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/af70f5e04c69839a1a0e08942254c170dc4c3d61)

malfet commented:
@pytorchbot revert -m 'Sorry for reverting your change but its tests are failing in trunk' -c nosignal

malfet commented:
@pytorchbot revert -m 'Take 2, of stack revert your change but its tests are failing in trunk' -c nosignal

jbschlosser committed:
Update on "Forward / backward NJT support for several activation functions"


Several activation functions were unimplemented due to missing `pointwise` tags. This PR adds them and corresponding backwards implementations.

[ghstack-poisoned]

jbschlosser committed:
Update on "Forward / backward NJT support for several activation functions"


Several activation functions were unimplemented due to missing `pointwise` tags. This PR adds them and corresponding backwards implementations.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Forward / backward NJT support for several activation functions"


Several activation functions were unimplemented due to missing `pointwise` tags. This PR adds them and corresponding backwards implementations.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Forward / backward NJT support for several activation functions"


Several activation functions were unimplemented due to missing `pointwise` tags. This PR adds them and corresponding backwards implementations.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge -i
